gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/jupiter/handlers/api/v4/AsyncApiReactor.java

```diff
@@ -192,14 +209,29 @@ private void setApiResponseTimeMetric(ExecutionContext ctx) {
 
     @Override
     public List<Acceptor<?>> acceptors() {
-        return api
+        final List<Acceptor<?>> httpAcceptors = api
             .getDefinition()
             .getListeners()
             .stream()
             .filter(listener -> ListenerType.HTTP == listener.getType())
             .flatMap(listener -> ((HttpListener) listener).getPaths().stream())
             .map(path -> new DefaultHttpAcceptor(path.getHost(), path.getPath(), this))
             .collect(Collectors.toList());
+
+        final List<DefaultSubscriptionAcceptor> subscriptionAcceptors = api
+            .getDefinition()
+            .getListeners()
+            .stream()
+            .filter(listener -> listener.getType().equals(ListenerType.SUBSCRIPTION))
+            .map(listener -> ((SubscriptionListener) listener))
+            .map(path -> new DefaultSubscriptionAcceptor(this, api.getId()))
+            .collect(Collectors.toList());
+
+        final List<Acceptor<?>> acceptors = new ArrayList<>();
```

Comment:
Could you try to build this list at once instead of browsing it twice ?

Reply:
Sometimes, the plain old fashioned way is by far the best one 😅
```java
    @Override
    public List<Acceptor<?>> acceptors() {
        final List<Acceptor<?>> acceptors = new ArrayList<>();

        for (Listener listener : api.getDefinition().getListeners()) {
            if (listener.getType() == ListenerType.HTTP) {
                acceptors.addAll(
                    ((HttpListener) listener).getPaths()
                        .stream()
                        .map(path -> new DefaultHttpAcceptor(path.getHost(), path.getPath(), this))
                        .collect(Collectors.toList())
                );
            } else if (listener.getType().equals(ListenerType.SUBSCRIPTION)) {
                acceptors.add(new DefaultSubscriptionAcceptor(this, api.getId()));
            }
        }

        return acceptors;
    }
```